### PR TITLE
[CUDA graphs] Add warning if captured graph is empty

### LINKED ISSUE #87894: [CUDAGraph] Silent failure when graphs capture attempted on wrong device
### 🐛 Describe the bug

The following code snippet silently fails to capture the graph due to memory being allocated on one device while the graphs capture being attempted on other.

```python
import torch

device = torch.device("cuda:0")
x = torch.randn(10, dtype=torch.float32, device=device)
y = torch.randn(10, dtype=torch.float32, device=device)
z = torch.zeros(10, dtype=torch.float32, device=device)

with torch.cuda.device('cuda:1'): # Wrong device
    g = torch.cuda.CUDAGraph()
    with torch.cuda.graph(g):
        z = x + y

    for i in range(3):
        x.normal_()
        y.normal_()
        g.replay()
        print(z) # One would expect it to print different values each iteration, 
                 # but it does not because the current_device is 0 
                 # while all the tensors are on device 1

    print(f'Test passed')
```

I believe it should raise an error. However, I couldn't think of any easy way of doing it. 
cc @ngimel @ptrblck 

### Versions

Latest

### PR DESCRIPTION
**Pull request by Aidyn-A**
Fixes #87894

This PR adds a warning if captured graph is empty (consists of zero nodes). 
The example snippet where would it be useful:

```python
import torch

x = torch.randn(10)
z = torch.zeros(10)

g = torch.cuda.CUDAGraph()
with torch.cuda.graph(g):
    z = x * x
# Warn user
```

and in #87894

### CODE REVIEW

File: test/test_cuda.py [hunk 1]
```diff
@@ -3259,6 +3260,16 @@ def test_graph_capture_simple(self):
...
 
+    @unittest.skipIf((not TEST_CUDA) or
+                     TEST_WITH_ROCM or
+                     int(torch.version.cuda.split(".")[0]) < 11, "CUDA >= 11.0 required for graphs")
+    def test_graph_warn_if_has_zero_nodes(self):
+        with warnings.catch_warnings(record=True) as caught:
+            g = torch.cuda.CUDAGraph()
+            g.capture_begin()
+            g.capture_end()
+        self.assertTrue(any("The CUDA Graph is empty" in str(w.message) for w in caught))
```
Comment (IvanYashchuk) [inline test/test_cuda.py:3271]: setting stream is missing here, but when added there's a weird interaction of streams and warnings. See:
```py
In [1]: import warnings

In [2]: import torch

In [3]: a = torch.randn(3, 3, device="cuda")

In [4]: out = torch.empty_like(a)

In [5]: with warnings.catch_warnings(record=True) as w:
   ...:     torch.linalg.cond(a, 2, out=out)
   ...:     print(w)
   ...:
[<warnings.WarningMessage object at 0x7ffbe05dce50>]

In [6]: with warnings.catch_warnings(record=True) as w:
   ...:     s = torch.cuda.Stream()
   ...:     with torch.cuda.stream(s):
   ...:         torch.linalg.cond(a, 2, out=out)
   ...:     print(w)
   ...:
[]
```

Comment (IvanYashchuk) [reply test/test_cuda.py:3271]: Example was incorrect.

File: test/test_cuda.py [hunk 2]
```diff
@@ -3259,6 +3260,18 @@ def test_graph_capture_simple(self):
 
         self.assertTrue(b.sum().item() == 11000.)
 
+    @unittest.skipIf((not TEST_CUDA) or
+                     TEST_WITH_ROCM or
+                     int(torch.version.cuda.split(".")[0]) < 11, "CUDA >= 11.0 required for graphs")
+    def test_graph_warn_if_has_zero_nodes(self):
+        with warnings.catch_warnings(record=True) as caught:
```
Comment (Aidyn-A) [inline test/test_cuda.py:3267]: The test is having issues with catching the warnings. @jjsjann123 pointed out that the possible cause is call_guard in pybing11  https://github.com/pytorch/pytorch/blob/master/torch/csrc/cuda/Graph.cpp#L31-L39 

Any ideas on how to catch this warning?
cc @ezyang @ngimel

Comment (ezyang) [reply test/test_cuda.py:3267]: You need the wrapper that clears the warning queue into Python. `HANDLE_TH_ERRORS` and `END_HANDLE_TH_ERRORS_PYBIND` check torch/csrc/README.md

cc @peterbell10

Comment (peterbell10) [reply test/test_cuda.py:3267]: I was going to suggest `torch::wrap_pybind_function` but it didn't work with member functions, so I've added that support and wrapped the CUDA graph functions in #88932. So, if you rebase it should start catching the warnings.

Comment (Aidyn-A) [reply test/test_cuda.py:3299]: Thanks @peterbell10, 
The test is passing!

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**